File: backend/src/main/java/com/drissman/service/AuthService.java
package com.drissman.service;

import com.drissman.api.dto.AuthResponse;
import com.drissman.api.dto.LoginRequest;
import com.drissman.api.dto.RegisterRequest;
import com.drissman.domain.entity.User;
import com.drissman.domain.repository.UserRepository;
import com.drissman.security.JwtTokenProvider;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;

    public Mono<AuthResponse> register(RegisterRequest request) {
        return userRepository.existsByEmail(request.getEmail())
                .flatMap(exists -> {
                    if (exists) {
                        return Mono.error(new RuntimeException("Email already exists"));
                    }

                    User user = User.builder()
                            .email(request.getEmail())
                            .password(passwordEncoder.encode(request.getPassword()))
                            .firstName(request.getFirstName())
                            .lastName(request.getLastName())
                            .role(User.Role.valueOf(request.getRole().toUpperCase()))
                            .build();

                    return userRepository.save(user)
                            .map(this::createAuthResponse);
                });
    }

    public Mono<AuthResponse> login(LoginRequest request) {
        return userRepository.findByEmail(request.getEmail())
                .switchIfEmpty(Mono.error(new RuntimeException("Invalid credentials")))
                .flatMap(user -> {
                    if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
                        return Mono.error(new RuntimeException("Invalid credentials"));
                    }
                    return Mono.just(createAuthResponse(user));
                });
    }

    private AuthResponse createAuthResponse(User user) {
        String token = jwtTokenProvider.generateToken(
                user.getId(),
                user.getEmail(),
                user.getRole().name());

        return AuthResponse.builder()
                .user(AuthResponse.UserDto.builder()
                        .id(user.getId())
                        .email(user.getEmail())
                        .firstName(user.getFirstName())
                        .lastName(user.getLastName())
                        .role(user.getRole().name())
                        .schoolId(user.getSchoolId())
                        .build())
                .token(token)
                .build();
    }
}

File: frontend/src/lib/auth-service.ts
import { apiClient } from "@/lib/api-client";

export interface AuthUser {
    id: string;
    email: string;
    firstName: string;
    lastName: string;
    role: "SCHOOL_ADMIN" | "CANDIDAT" | "MONITOR";
    schoolId?: string;
}

export interface AuthResponse {
    token: string;
    user: AuthUser;
}

export interface RegisterPayload {
    email: string;
    password: string;
    firstName: string;
    lastName: string;
    phone?: string;
    role: "CANDIDAT" | "SCHOOL_ADMIN";
    schoolName?: string;
}

export interface LoginPayload {
    email: string;
    password: string;
}

export const authService = {
    register: (payload: RegisterPayload) =>
        apiClient.post<AuthResponse>("/auth/register", payload),

    login: (payload: LoginPayload) =>
        apiClient.post<AuthResponse>("/auth/login", payload),
};
